File: internal/chat/message.go
package chat

type MessageType string
type Message struct {
	ID       ID          `json:"id"`
	RoomID   ID          `json:"roomID"`
	ClientID ID          `json:"clientID,omitempty"`
	UserID   ID          `json:"userID,omitempty"`
	UserName string      `json:"userName,omitempty"`
	Type     MessageType `json:"type"`
	To       []ID        `json:"to,omitempty"`
	Omit     []ID        `json:"omit,omitempty"`
	Refetch  []string    `json:"refetch,omitempty"`
	Text     string      `json:"text,omitempty"`
	Payload  any         `json:"payload,omitempty"`
}

const (
	System        MessageType = "system"
	SystemChat    MessageType = "system:chat"
	Connect       MessageType = "connect"
	Chat          MessageType = "chat"
	DirectMessage MessageType = "direct_message"
)

File: web/src/types/message.ts
import type { RoomClient } from './room'

export type ConnectMessage = {
  id: string
  type: 'connect'
  roomID: string
  refetch: string[]
  payload: RoomClient

  timestamp?: Date
}

export type SystemMessage = {
  id: string
  type: 'system'
  roomID: string
  refetch?: string[]
  text?: string
  payload?: {
    client: RoomClient
  }

  timestamp?: Date
}

export type ChatMessage = {
  id: string
  type: 'chat'
  roomID: string
  clientID: string
  userID: string
  userName: string
  text: string

  timestamp?: Date
}

export type SystemChatMessage = {
  id: string
  type: 'system:chat'
  roomID: string
  clientID?: string
  refetch?: string[]
  text: string
}

export type Message =
  | ConnectMessage
  | SystemMessage
  | ChatMessage
  | SystemChatMessage
export type ClientMessage = ChatMessage
